### ISSUE TITLE
adding a record to an ontology

### ISSUE THREAD

**Issue by jkobject** [laminlabs/lamindb#1308]
Hi, Thanks again for the great tool!

I am using ontologies and very much like the way to use `$ontology.validate()`, `$ontology.df()`, etc.. which I get by doing something like this `$ontology = lb.$ontology.bionty()`, `$ontology` can be celltype, tissues, etc..

I am using these default ontologies that contain most of what I need. 
However, I often need to add a couple missing values to those. 
You give an example where someone can add a new record to `lb.$ontology` but how do I get it merged with the ontology from `lb.$ontology.bionty()` so that when I now do .validate(), .df(), etc. it is also using my new value?

**Comment by sunnyosun** [comment]
Hi @jkobject,

While `bionty` is used to manage public ontologies, managing custom ontology terms can be done with LaminDB registries (via `lnschema-bionty`). You can create records in your LaminDB registries from the public ontologies or from your custom terms. `.validate` is automatically performed against existing records in your instance.

Note that the bionty accessor `.bionty()` is only aware of the public ontologies, but you have all the same methods in LaminDB. e.g. `lb.CellType.bionty().validate` vs `lb.CellType.validate`.

To see detailed examples, feel free to check out https://lamin.ai/docs/bio-registries.

**Comment by falexwolf** [comment]
Thanks, Sunny!

Jeremie, how can we more clearly communicate that `lb.CellType.validate()` etc. exists in the docs? Or, where did your impression come from that these things only exist on the `Bionty` objects?

As Sunny puts it, `Bionty` is a mere way of unifying access to public ontologies, but typically, you'd work with the ontology clones in your own in-house ontologies.

You see that the concept of transferring data from public knowledge and other lamindb instances is meant to give you the freedom to iterate on your own problem without requiring you to re-curate data & metadata.

I just fear we struggle to communicate this concept well - the only analogy that people know are git clones and forks, and these are related but also quite different when it comes to specifics.

**Comment by jkobject** [comment]
Hi, 

I see that one can do `validate` from lb.CellType but if for example I want to do `df()`, this doesn't work...
Let's say I want my ontology to be everything in bionty + some cell types of mine, I would want then to be able to do everything I could do from the bionty version. E.g. `validate()`, `df()`, `inspect()`...

**Comment by jkobject** [comment]
Hi, 

I see that one can do `validate` from lb.CellType (it wasn't obvious since lb.CellType doesn't behave the same as the object that lb.Celltype() returns) 

but if for example I want to do `lb.CellType.df()`, this doesn't exists...

If I want my ontology to be everything in bionty + some cell types of mine, how would I do that?
I would want then to be able to do everything I could do from the bionty ontology version. E.g. `validate()`, `df()`, `inspect()`...

**Comment by sunnyosun** [comment]
If you want to add everything from Bionty to your registry table, you can use `.from_values([list_of_ontology_ids], field="ontology_id")` to create all records from bionty, and then save them in your own registry.

Once you've done that, you can use `lb.CellType.filter().df` to see the entire table, similarly you have `lb.CellType.

Example code:
```
bionty_df = bionty.CellType().df()
records = lb.CellType.from_values(bionty_df.index, field=lb.CellType.ontology_id)
ln.save(records) # this will take some time, due to the hierachies

# add some of your own records
lb.CellType(name=...).save()

# afterwards, you have a combined registry, and can do:
```
lb.CellType.validte()
lb.CellType.inspect()
lb.CellType.standardize()
lb.CellType.filter().df
```

**Comment by jkobject** [comment]
Oh ok I understand now how this works!

It feels like the same issue as for the datasets. :)  

So it would be nice to have something that makes it clear that there is a "local" and a "remote" version

Like `lb.CellType` is the local and `lb.CellType.bionty()` is the remote (or even better if the remote is only accessible through `bionty`

Then for both of those to behave as similar as possible and it is just a choice to work with an "unmodifiable" ground truth remote or with a "personal" local

Just an idea

**Comment by sunnyosun** [comment]
Thank you for the suggestion @jkobject! It's a good idea and we'll certainly improve the clarity for this since it led to confusion for you.

Yes, the general idea is that `Bionty` is the accessor to the public knowledge and LaminDB registry is your in-house knowledge. You can import public knowledge by seeding your registry with knowledge-coupled records. But of course the main point is that you can maintain your own registry by adding and modifying records there. The public knowledge is always tracked with a version, so you will always know how those records are created.

**Comment by jkobject** [comment]
Just so you know, this worked on my end but:

I got an error that appeared for big ontologies: "OperationalError: database is locked"

Also, now in my personal ontology version, I don't see any parent relationships..

for any indices: `lb.CellType.lookup()[1].parents.df()` returns an empty df
and `lb.CellType.filter().df()` doesn't have a column "parent"

**Comment by sunnyosun** [comment]
Which type of database are you using? Are you using a remote SQLite? If so, when the database is locked, it means that other users are currently connecting to it. (To resolve, they need to run "lamin close" to unlock the instance for you.)

If you don't see parents, perhaps you've set `auto_save_parents=False`? The default should be True and you should see a warning message about adding parents when you run `ln.save(records)`. If you've already added records without any parents, you need to clean them up with `ln.CellType.filter().delete()` and re-add them again.

Note `lb.CellType.filter().df()` by default doesn't show ManyToMany relationship columns includes "parents", if you want to see that, you need to call something like `lb.CellType.filter().df(include=["parents__name"])`.

**Comment by jkobject** [comment]
I understand. I think the issue is that I was running multiple notebooks in parallel, maybe one of them was accessing the database at the same time.
Doing: lb.CellType.filter().df(include=["parents__ontology_id"]) did show what I wanted. Thanks!

**Comment by falexwolf** [comment]
> I think the issue is that I was running multiple notebooks in parallel, maybe one of them was accessing the database at the same time.

Hmm, this should be an issue. Your SQLite database is multi-threaded and you should have no problems writing to the same database from a bunch of notebooks at the same time.

Please let us know if the error occurs again with a traceback!

**Comment by falexwolf** [comment]
In lamindb 0.62, we now added `.df()` to `Registry` so that you'll get the consistent behavior you were missing on this front.

This had been requested on the team for a while.

PR: https://github.com/laminlabs/lamindb/pull/1313

**Comment by jkobject** [comment]
Maybe still following the same schema of the bionty package. Could it be possible to easily select the organism in the lb field by writing something like this:
`lb.DevelopmentalStage(organism="mouse")` ?

**Comment by sunnyosun** [comment]
> Maybe still following the same schema of the bionty package. Could it be possible to easily select the organism in the lb field by writing something like this: `lb.DevelopmentalStage(organism="mouse")` ?

Currently, only `Gene`, `Protein`, and `CellMarker` are organism-aware, as multiple organisms can come from the same source. For `DevelopmentalStage`, it has a one-to-one correspondence of source and organism, and therefore, in this case, you need to pass bionty_source like so:

```
bionty_source_ds_mouse = lb.BiontySource.filter(entity="DevelopmentalStage", organism="mouse").one()
records = lb.DevelopmentalStage.from_values(names, field=lb.DevelopmentalStage.ontology_id, bionty_source=bionty_source_ds_mouse)
ln.save(records)
```

**Comment by jkobject** [comment]
I understand. And would you know why changing the `settings.organism` does not change the `lb.Gene.df()` values? I still get human by default
